### PR TITLE
Add AVS rule to Fraud & Risk tools

### PR DESCRIPTION
**Pull request by eduardoumpierre**
Fixes Automattic/woocommerce-payments-server#3346

#### Changes proposed in this Pull Request

This PR aims to add the AVS rule to the advanced settings page of the Fraud & Risk tools.

#### Testing instructions

* Checkout to this branch
* Listen to Stripe webhooks
* Go to **Payments > Settings**
* Scroll down to the **Fraud protection** section, select "Advanced" and click on "Configure" (or "Edit" if it has already been configured)
* The toggle on the rule cards means that now if it should be blocked, so there is no "Review" option – make sure the UI only mentions the payment being blocked
* Search for the "AVS mismatch" rule and toggle it, then save
* Access [Stripe's dashboard](https://dashboard.stripe.com/test/logs?showIP=false&created[preset]=1H)
  * Check the latest `POST /v1/accounts/YOUR_ACCOUNT_ID`
  * Scroll the right container to the bottom of the page and you should see a similar **Request POST body**

image

-------------------

- [x] Run `npm run changelog` to add a changelog file, choose `patch` to leave it empty if the change is not significant. You can add multiple changelog files in one PR by running this command a few times. 
- [x] Covered with tests (or have a good reason not to test in description ☝️)
- [ ] Tested on mobile (or does not apply)

**Post merge**

- [ ] Link to testing instructions from [release testing doc](https://github.com/Automattic/woocommerce-payments/wiki/Release-testing-instructions) following [these instructions](https://github.com/Automattic/woocommerce-payments/wiki/How-to-write-good-manual-testing-scenarios) : _Add link here / 'QA Testing Not Applicable'_
- [ ] Add or update [critical flows](https://github.com/Automattic/woocommerce-payments/wiki/Critical-flows) and [testing instructions for critical flows](https://github.com/Automattic/woocommerce-payments/wiki/Testing-instructions-for-critical-flows), if applicable.
- [ ] Add what's changed (description, screenshot, demo videos etc.) to the release announcement post, if applicable.

### CODE REVIEW

**Comment by botwoo** [comment]
### Test the build 

#### Option 1. Jetpack Beta

- Install and activate [Jetpack Beta](https://jetpack.com/download-jetpack-beta/).
- Use this build by searching for PR number `6275` or branch name `add/3346-avs-checks` in your-test.site/wp-admin/admin.php?page=jetpack-beta&plugin=woocommerce-payments          

#### Option 2. Jurassic Ninja - available for logged-in A12s

:rocket: [Launch a JN site with this branch](https://jurassic.ninja/create/?jetpack-beta&shortlived&nojetpack&woocommerce&branches.woocommerce-payments=add/3346-avs-checks) :rocket:

:information_source: Install this [Tampermonkey script](https://github.com/Automattic/woocommerce-payments/tree/develop/bin/wcpay-live-branches) to get more options.

---

Build info: 

- Latest commit: b5cc4cb036bd1fa5d7de8affe8aa2dd8def455dd
- Build time: 2023-05-14 19:09:13 UTC

_Note: the build is updated when a new commit is pushed to this PR._

**Review by allie500 — COMMENTED** [review]
Looks like an unused import in client/settings/fraud-protection/advanced-settings/test/rule-toggle.test.tsx - `userEvent`.

File: client/globals.d.ts [hunk 1]
```diff
@@ -89,3 +90,5 @@ declare global {
 
 	const wcSettings: Record< string, any >;
 }
+
+declare const wcTracks: any;
```
Comment (dwainm) [inline client/globals.d.ts]: Let's see why this was added as I didn't have it in my PR.

File: client/settings/fraud-protection/advanced-settings/rule-toggle.tsx [hunk 2]
```diff
@@ -11,12 +11,14 @@ import { ToggleControl, RadioControl } from '@wordpress/components';
 import './../style.scss';
 import FraudPreventionSettingsContext from './context';
 
+const isReviewEnabled = false;
```
Comment (dwainm) [inline client/settings/fraud-protection/advanced-settings/rule-toggle.tsx]: We should change this to use the feature flag.

File: client/settings/fraud-protection/advanced-settings/rule-toggle.tsx [hunk 3]
```diff
@@ -33,16 +35,17 @@ const radioOptions = [
 ];
 
 const helpTextMapping = {
-	unchecked: __(
-		'When enabled, the payment method will not be charged until you review and approve the transaction.'
-	),
+	unchecked: __( 'When enabled, the payment method will be blocked.' ),
```
Comment (dwainm) [inline client/settings/fraud-protection/advanced-settings/rule-toggle.tsx]: The payment will be blocked, not the method.

File: client/settings/fraud-protection/advanced-settings/test/rule-toggle.test.tsx [hunk 4]
```diff
@@ -155,29 +146,4 @@ describe( 'Fraud protection rule toggle tests', () => {
 			mockContext.protectionSettingsUI.test_rule.enabled
 		).toBeFalsy();
 	} );
-	test( 'sets the value correctly when block is selected', () => {
```
Comment (dwainm) [inline client/settings/fraud-protection/advanced-settings/test/rule-toggle.test.tsx:158]: Rather skip these tests so we can reuse them later.

File: client/transactions/test/index.tsx [hunk 5]
```diff
@@ -205,22 +205,5 @@ describe( 'TransactionsPage', () => {
 
 		await renderTransactionsPage();
 		expect( screen.queryByText( /blocked/i ) ).toBeInTheDocument();
-		expect( screen.queryByText( /risk review/i ) ).toBeInTheDocument();
-	} );
-
-	test( 'do not render fraud outcome tabs if the feature flag is disabled', async () => {
```
Comment (dwainm) [inline client/transactions/test/index.tsx:211]: Skip this one as well.

File: includes/class-wc-payments-account.php [hunk 6]
```diff
@@ -1322,6 +1322,7 @@ private function settings_changed( $changes = [] ) {
 		}
 
 		$diff = array_diff_assoc( $changes, $account );
+
```
Comment (dwainm) [inline includes/class-wc-payments-account.php]: uneeded change

**Review by dwainm — APPROVED** [review]

**Comment by dwainm** [comment]
Some minor changes were requested, but over all it looks good and tests as advertised.

File: client/settings/fraud-protection/advanced-settings/test/rule-toggle.test.tsx (hunk 4 above)
Comment (eduardoumpierre) [reply client/settings/fraud-protection/advanced-settings/test/rule-toggle.test.tsx:158]: Fixed on [89e8fb8](https://github.com/Automattic/woocommerce-payments/pull/6275/commits/89e8fb8a7ca1bf8b090c0627a290fa60c973135b).

File: client/transactions/test/index.tsx (hunk 5 above)
Comment (eduardoumpierre) [reply client/transactions/test/index.tsx:211]: Fixed on [89e8fb8](https://github.com/Automattic/woocommerce-payments/pull/6275/commits/89e8fb8a7ca1bf8b090c0627a290fa60c973135b).

File: includes/class-wc-payments-account.php (hunk 6 above)
Comment (eduardoumpierre) [reply includes/class-wc-payments-account.php]: Removed on [815bd03](https://github.com/Automattic/woocommerce-payments/pull/6275/commits/815bd03f1153b5d01fa6c03f78edd9b47fac1fc7).

File: client/globals.d.ts (hunk 1 above)
Comment (eduardoumpierre) [reply client/globals.d.ts]: Removed on [815bd03](https://github.com/Automattic/woocommerce-payments/pull/6275/commits/815bd03f1153b5d01fa6c03f78edd9b47fac1fc7). It was probably a leftover from a merge conflict resolution. Thanks for catching it!

File: client/settings/fraud-protection/advanced-settings/rule-toggle.tsx (hunk 3 above)
Comment (eduardoumpierre) [reply client/settings/fraud-protection/advanced-settings/rule-toggle.tsx]: Fixed on [43ba3f7](https://github.com/Automattic/woocommerce-payments/pull/6275/commits/43ba3f7b7d22b12fd1be6d6913638671b90280be).

File: client/settings/fraud-protection/advanced-settings/rule-toggle.tsx (hunk 2 above)
Comment (eduardoumpierre) [reply client/settings/fraud-protection/advanced-settings/rule-toggle.tsx]: Updated on [98e56e0](https://github.com/Automattic/woocommerce-payments/pull/6275/commits/98e56e0d36d5ac17d9937b95f57ce8c0a74c269b)!

File: includes/admin/class-wc-rest-payments-settings-controller.php [hunk 7]
```diff
@@ -799,10 +799,34 @@ private function update_fraud_protection_settings( WP_REST_Request $request ) {
 		$this->api_client->save_fraud_ruleset( $ruleset_config );
 
 		// Update local cache.
+		$this->wcpay_gateway->update_cached_account_data(
```
Comment (tpaksu) [inline includes/admin/class-wc-rest-payments-settings-controller.php:802]: This assumes that we always are successful in terms of setting the AVS check flag. But we're not always successful in setting this flag. For example in a test account, that may cause confusion. Just leaving the note here to replace that after all required features are released that will allow us to remove the platform level AVS check.

File: client/transactions/test/index.tsx (hunk 5 above)
Comment (tpaksu) [reply client/transactions/test/index.tsx:211]: Maybe we should leave a todo comment not to forget it.

**Review by tpaksu — APPROVED** [review]
All works as described in the PR description. The AVS activation and deactivation is fast, may help us in the future to leave some TODO comments to remove the code added on this PR that will be removed when server side confirmation is merged. LGTM, we can ship it.

**Pull request merged**